src/transport/MessageHeader.cpp

```diff
@@ -147,9 +154,9 @@ CHIP_ERROR PacketHeader::Decode(const uint8_t * data, size_t size, size_t * deco
 
     if (mFlags.value & Header::Flags::kSourceNodeIdPresent)
     {
-        VerifyOrExit(size >= kNodeIdSizeBytes, err = CHIP_ERROR_INVALID_ARGUMENT);
+        static_assert(kNodeIdSizeBytes == 8, "Read64 might read more bytes than we have in the buffer");
+        VerifyOrExit(size >= (p - data) + kNodeIdSizeBytes, err = CHIP_ERROR_INVALID_ARGUMENT);
```

Comment:
Does this approach save any storage/performance?

It sems to slightly lose readability as `size >= (p - data) + kNodeIdSizeBytes` seems harder to parse than `size >= kNodeIdSizeBytes`

Reply:
> Does this approach save any storage/performance?

No, but it facilitates correctness.  The existing code was buggy.  It did the following:

1. Check that size is `>= kFixedUnencryptedHeaderSizeBytes`
2. Read some data.
3. Check that size is `>= kNodeIdSizeBytes` (if the relevant flag is set)

Note the lack of and "update `size`" step between steps 2 and 3.  This could lead to us reading off the end of the buffer.  (And I'm surprised we have no test for this... I should add one, now that I think about it.)

That could be fixed by updating `size`, but the number of bytes the code reads in step 1 is _not_ `kFixedUnencryptedHeaderSizeBytes`, so the `size` update would require some non-obvious work.  That can be done, of course...

With the approach I took, we check the actual thing we care about: given the current read position and the known total size of the buffer, can we read the number of bytes we plan to read?  There's no way to forget or fail to keep `size` in sync with the position in this approach.  I agree that it would be nice to make it more readable somehow.

Maybe what we should do is have variants of the `LittleEndian::Read*` methods that take both a position reference and a size reference and update both, so the synchronization is centralized.  If we did that we could just do tests that `size` is big enough.

Reply:
Per offline discussion, filed https://github.com/project-chip/connectedhomeip/issues/2914 as a followup to make this nicer.